text/0069-api-v2.md

```diff
@@ -0,0 +1,194 @@
+
+1. RawKV is not safe to use along with TxnKV. By solving this, TiDB will be able to support RawKV as the table's storage engine, which will enrich TiDB's use cases.
+2. RawKV TTL is controlled by Store configuration. Switching the configuration will cause data corruption in silence.
+3. RawKV TTL is encoded into the value by appending 8-bytes UNIX timestamp to the end of the value, therefore it's hard to introduce other encode afterward.
+4. RawKV compare_and_swap is easy to make mistake.
+5. It could be nice if we can deploy multiple applications on one TiKV cluster.
+
+## Detailed Design
+
+### New key-value codec
+
+This RFC introduces a new key encode to RawKV and TxnKV, and a new value encode to RawKV, which will allow the RawKV to be used along with TxnKV and also allow TiKV to flexibly add meta, e.g. the TTL, to a RawKV value.
+
+In addition, keys will be contained in keyspaces, where the keys in different keyspace are totally independant. If keyspace is not specified, the keyspace 'default' will be used.
+
+The `API V2` is enabled by a switch on PD. Since it changes the storage encode, it will be not compatible to switch between `API V1` and `API V2` while there are non-TiDB data in TiKV. TiDB data is specially treated in order to not be affected by the change.
+
+#### Key Encode
+
+Once the `API V2` is enabled, the key will be starting with either:
+
+1. `m` or `t`: TxnKV key. Used by TiDB.
+2. `k{keyspace prefix id}x`: TxnKV key.
+3. `k{keyspace prefix id}r`: RawKV key.
+4. `k{keyspace prefix id}a`: RawKV atomic key, on which you can use `CompareAndSwap`.
+
+The `{keyspace prefix id}` is the [keyspace](https://github.com/tikv/rfcs/pull/39) prefix for seperating keys of different keyspace. It should be an vary-length integer whose highest bit of every byte denotes whether the next byte is still part of the integer. The client will fetch the prefix from PD by the keyspace name specified by the user when initializing the client, so it means that the keyspace prefix is valid during the session, in other words, the change on keyspace on PD will not take effect on running seesions.
+
+Note that in TxnKV, the key will be encoded by `Memory Comparable Encoding`. But since the `Memory Comparable Encoding` will not change the starting bytes but only add paddings, there won't be any overlap between RawKV and TxnKV.
+
+#### RawKV Value Encode
+
+If the key has RawKV prefix, then the value can be either:
+
+1. `{0x0}{data}`
+2. `{0x1}{TTL expire timestamp}{data}`
+
+#### TTL in Titan
+
+Titan is a storage engine plugin for RocksDB that store the value of a kvpair in a separated blob file if the value is large enough. This brings a new challenge for RawKV TTL compaction filter where the compaction filter can not know the ttl without reading the blob file. So with Titan, the RawKV value encode is defined to be:
+
+##### Value
+
+1. `{0x0}`
+2. `{0x1}{TTL expire timestamp}`
```

Comment:
Is it possible to have other encoding schema later on? If not, we can probably use varint encoding to encode TTL lifetime and assume lifetime of 0 means no expiration. This way values are encoded  exactly the same way.

Reply:
It would be nice to have TTL support in TxnKV.

Reply:
> Is it possible to have other encoding schema later on.

Yes, it's possible, just to use a new prefix byte in the value to indicate the new format.

Reply:
Just have a second thought, how about we make this encoding works for V2 TxnKV?